rust/gui-client/src-tauri/src/client/tunnel_wrapper/windows.rs

```diff
@@ -0,0 +1,109 @@
+use anyhow::{anyhow, Context as _, Result};
+use connlib_client_shared::Callbacks;
+use firezone_headless_client::{IpcClientMsg, IpcServerMsg};
+use futures::{SinkExt, Stream};
+use secrecy::{ExposeSecret, SecretString};
+use std::{pin::pin, task::Poll};
+use tokio::{net::windows::named_pipe, sync::mpsc};
+use tokio_util::codec::{Framed, LengthDelimitedCodec};
+
+pub(crate) struct TunnelWrapper {
+    task: tokio::task::JoinHandle<Result<()>>,
+    // Needed temporarily to avoid a big refactor. We can remove this in the future.
+    tx: mpsc::Sender<String>,
+}
+
+impl TunnelWrapper {
+    pub(crate) async fn disconnect(mut self) -> Result<()> {
+        self.send_msg(&IpcClientMsg::Disconnect)
+            .await
+            .context("Couldn't send Disconnect")?;
+        self.task.abort();
+        Ok(())
+    }
+
+    #[allow(clippy::unused_async)]
+    pub(crate) async fn send_msg(&mut self, msg: &IpcClientMsg) -> Result<()> {
+        self.tx
+            .send(serde_json::to_string(msg).context("Couldn't encode IPC message as JSON")?)
+            .await
+            .context("Couldn't send IPC message")?;
+        Ok(())
+    }
+}
+
+enum IpcEvent {
+    /// The client wants to send a message to the service
+    Client(String),
+    /// The connlib instance in the server wants to send a message to the client
+    Connlib(IpcServerMsg),
+}
+
+pub(crate) async fn connect(
+    api_url: &str,
+    token: SecretString,
+    callback_handler: super::CallbackHandler,
+    tokio_handle: tokio::runtime::Handle,
+) -> Result<TunnelWrapper> {
+    tracing::info!(pid = std::process::id(), "Connecting to IPC service...");
+    let ipc = named_pipe::ClientOptions::new()
+        .open(firezone_headless_client::windows::pipe_path())
+        .context("Couldn't connect to named pipe server")?;
+    let ipc = Framed::new(ipc, LengthDelimitedCodec::new());
+    // This channel allows us to communicate with the GUI even though NamedPipeClient
+    // doesn't have `into_split`.
+    let (tx, mut rx) = mpsc::channel(1);
+
+    let task = tokio_handle.spawn(async move {
+        let mut ipc = pin!(ipc);
+        loop {
+            let ev = std::future::poll_fn(|cx| {
```

Comment:
Is this the bit you were looking to get @thomaseizinger's input on?

Reply:
As with any eventloop architecture I guess the one thing to double-check here is that processing of messages doesn't get held up.

Since these are control messages and not hot path I think we should be fine, but just calling out for a second look.

Reply:
Yeah, I'm trying to imitate that poll style he uses, to avoid Tokio's `select` and having extra tasks.
With that style you can't await inside the poll function, so in one or two cases I made an enum and use the poll function to wait on the next event, then process the event in a regular async block.

So processing does wait on things. If we need to break out async tasks we can do that, it's just more complex

Reply:
why do we want to avoid `select`?

Reply:
Thomas said so 🤷‍♀️ 
I'm guessing because it uses randomness by default and it's hard to tell whether any given function is cancel safe. There is one function in Tokio which has its cancel safety documented and I still cannot tell whether it's cancel safe.

Reply:
Comment about cancel-safety: https://github.com/firezone/firezone/pull/3394#issue-2099368578

Comment on one of my recent PRs: https://github.com/firezone/firezone/pull/4198#discussion_r1538393108

Reply:
Marking this off, #5158 will do this https://github.com/firezone/firezone/pull/5158/files#diff-f6659ed6aa430d8a07a8f29d236674a81a15ac1c853cd2b8a1dbfc567693c828R153